### ISSUE TITLE
Implementing short-sell option by altering data-feed

### ISSUE THREAD

**Issue by ALevitskyy** [ZhengyaoJiang/PGPortfolio#95]
I am not sure whether this is a right place to ask my question, but the paper seems to use a long-only strategy, which has its drawbacks in bear market. I want to add short-sells,  and not being able to change inner-working of the model to support this feature and not willing to debug any side-effects of my changes, I want to achieve that by adding extra instruments to data-feed which would move in opposite directions to originals. So for example for a time series X_{t} I would do something like Y_{t}=-X_{t}+2*MEAN(X), where adding 2*MEAN(X) avoids negative values (I am not sure whether the framework supports negatives). I will implement my own back-testing strategy, so backtesting with fake quotes should not be a problem. Is there any way this approach may have negative side effects, or is this a valid solution? I also was thinking of implementing leverage by doing something like Y_{t}=leverage*X_{t} is this a bad idea too?

**Comment by DexHunter** [comment]
I think shorting is more about actions. Can you explain a bit more on why do you think changing the input data would help the agent's decision?

**Comment by ALevitskyy** [comment]
Shorting is like buying an asset which moves in opposite direction to original, at least as a first-order approximation. So what I am doing now is just altering Data.db by creating extra asset, rETH which moves in opposite direction to ETH using the formula rETH=1/ETH. So the agent can chose this asset if it wants to short rather than go long on a position. The problem for me for now is that the trained agent assigns some portion of portfolio to long and short at the same time which is a bit counter-intuitive, hence I am asking whether it is a legitimate approach to introduce another "reverse" asset as an option.

**Comment by davidsblom** [comment]
I guess a different approach to incorporate short trades (and possibly also leverage), is to change the loss function and the output of the network. 

I'd say that you have three outputs per asset: the first output is a boolean (or zero/one) indicating long or short position, a second value which is the portfolio weight for the long position, and a third value for the short position. You would need to write a custom softmax function.
The loss function will only add the contribution of either the short or the long trade.

**Comment by DexHunter** [comment]
Yes, what David wrote is a good method. Besides, you can change the
activation function to allow negative output.

**Comment by davidsblom** [comment]
Yeah, you could also use the same variable to control long/short position. I think you should use the tanh activation function, and then an L1 normalization step in the output layer to allow for short trades.

Do you still need to change the loss function? I don't think so, right?

**Comment by DexHunter** [comment]
> Do you still need to change the loss function? I don't think so, right?

I think you still need to change the loss function. Besides tanh, you can try other activation functions ( [full list at wikipedia](https://en.wikipedia.org/wiki/Activation_function#Comparison_of_activation_functions) )

**Comment by kenuku** [comment]
This is wrong approach to calculate rETH=1/ETH (or to be precise rETHBTC=1/ETHBTC)
1/ETHBTC means price BTC in ETH and equivalently BTCETH
1/ETHBTC is NOT equivalent shorting price of ETHBTC

**Comment by jgerlach93** [comment]
hey, dont know if anyone is still interested in this topic. 
but shouldnt it be possible to simply replace the softmax layer at the end of the network by a different operation on the weight vector? 
i am not sure about what values the voting layer returns. are they positive only and then normalized by the softmax, or can they be negative as well? because if the latter is true, one could simply take the "raw" output vector from the voting layer, containing positive and negative values, and set the final layer as w' = w/sum(abs(w)). this will return a weight vector w' with positive and negative values for long and short, respectively, whose absolute values sum(abs(w'))=1 always. then, the network should learn to make use of negative expected growth, as well, right? or is my understanding of the voting layer wrong, and it is more a "counting" of positive-valued votes from the individual constituents of the ensemble?

**Comment by trevorprater** [comment]
I think the above comment makes a lot of sense, assuming the softmax layer is not performing a necessary function that was omitted above. This would be quite useful, receiving the weighted vector. Ideal for convex portfolio optimization, factor neutrality, and proper hedging, instead of, say, 1/USDT being the only _hedge_.

N.B. This repo is now quite old, all things considered. If I remember correctly (as being the third star), this came out long before futures markets were relevant or mature in the crypto space, leaving 1/USDT to be the only available, given that all pairs, at the time, used BTC as the underlier. In any case, yes, non-positive alpha signals would likely improve my trading pipeline.

**Comment by jgerlach93** [comment]
thanks for the response. yes, the repo is quite old, may be a good starting point, though. concerning the normalization, i implemented it by replacing the softmax function with my formula above. however, it is not quite as easy as i thought to get the whole thing right. there are several things: in my previous comment, i was ignorant of the fact that negative elements in wt will lead to wrong computation of the portfolio value, because negative returns are not correctly propagated by the implemented expressions, leading to negative arguments of the logarithms etc; consider for example the case where pt = 0.99, pt-1 = 1. then, yt = 0.99. if we had fully shorted at t-1 (i.e. wt-1 = -1), we would have (0.99*-1)-1=-1.99...
>>>> my point is, one will also have to adapt the loss function and a bit of the computation of the involved quantities such as the log-returns, in order to make it work. but then, it should be ok..
